File: lab2/triangle.py
from OpenGL.GL import *

import math
import utils

dimensions = utils.Coord(400, 400)
bounds = utils.Coord(100, 100)
viewport = utils.Viewport(bounds)


def startup():
    glClearColor(0.5, 0.5, 0.5, 0.1)
    viewport(None, dimensions)


def shutdown():
    pass


def make_triangle(side):
    height = side * math.sqrt(3) / 2

    glBegin(GL_TRIANGLES)
    glColor3f(1.0, 0.0, 0.0)
    glVertex2f(-side / 2, -height / 2)

    glColor3f(0.0, 1.0, 0.0)
    glVertex2f(side / 2, -height / 2)

    glColor3f(0.0, 0.0, 1.0)
    glVertex2f(0.0, height / 2)
    glEnd()


def render(time):
    glClear(GL_COLOR_BUFFER_BIT)
    make_triangle(200)
    glFlush()


def main():
    utils.default_init(dimensions=dimensions,
                       startup=startup,
                       render=render,
                       shutdown=shutdown,
                       viewport=viewport)


if __name__ == '__main__':
    main()

File: lab2/Triangle.cpp
#include <GLFW/glfw3.h>
#include <numbers>

#include "Triangle.h"

namespace gk
{

auto Triangle::run(Vector2i dimensions, bool vsync) -> void
{
    this->dimensions = dimensions;
    utils::configureViewport(bounds, nearDepth, farDepth);
    utils::defaultInit(dimensions,
                       [this] { startup(); },
                       [this](auto time) { render(time); },
                       [this] { shutdown(); },
                       vsync);
}

auto Triangle::startup() -> void
{
    glClearColor(0.5f, 0.5f, 0.5f, 0.1f);
    utils::viewport(nullptr, dimensions.x, dimensions.y);
}

auto Triangle::shutdown() -> void
{

}

auto Triangle::render([[maybe_unused]] double time) -> void
{
    glClear(GL_COLOR_BUFFER_BIT);
    makeTriangle(200.f);
    glFlush();
}

auto Triangle::makeTriangle(float side) -> void
{
    const auto height = side * std::numbers::sqrt3_v<float> / 2.f;

    glBegin(GL_TRIANGLES);
    glColor3f(1.f, 0.f, 0.f);
    glVertex2f(-side / 2.f, -height / 2.f);

    glColor3f(0.f, 1.f, 0.f);
    glVertex2f(side / 2.f, -height / 2.f);

    glColor3f(0.f, 0.f, 1.f);
    glVertex2f(0.f, height / 2.f);
    glEnd();
}

}

